kissh.py

```diff
@@ -0,0 +1,142 @@
+#!/usr/bin/env python3
+import argparse
+import pwd
+import subprocess
+import sys
+import traceback
+from pathlib import Path
+
+import requests
+
+session = requests.Session()
+
+
+def user_exists(user):
+    try:
+        user_info = pwd.getpwnam(user)
+        return Path(user_info.pw_dir) / ".ssh/authorized_keys"
+    except KeyError:
+        return None
+
+
+def create_user(user):
+    subprocess.run(["adduser", user, "--gecos=''", "--disabled-password"], check=True)
+    subprocess.run(["adduser", user, "sudo"], check=True)
+    # force password set on first ssh login, by deleting and expiring their
+    # passwd
+    subprocess.run(["passwd", "-de", user], check=True)
+
+    return user_exists(user)
+
+
+def get_fingerprint(key):
+    ps = subprocess.run(
+        ["ssh-keygen", "-lf-"],
+        input=key,
+        check=True,
+        capture_output=True,
+        text=True,
+    )
+    # ssh-keygen output format is: "SIZE HASH:VALUE no comment (TYPE)".
+    # eg: "256 SHA256:AsFa6mIf22oiMZSW7yNn3Fip2Ri6DAzjrh+KiZ+axWg no comment (ED25519)"
+    return ps.stdout.split()[1]
+
+
+def get_github_key(user, fingerprint):
+    """Get Github user's key that matches fingerprint."""
+    # Do not use the api to avoid ratelimiting
+    r = session.get(f"https://github.com/{user}.keys")
+    r.raise_for_status()
+
+    # Find the correct key by its fingerprint
+    for key in r.text.splitlines():
+        if get_fingerprint(key) == fingerprint:
+            return key
+
+
+def write_file(path, contents):
+    path.parent.mkdir(exist_ok=True)
+    path.write_text(contents)
+
+
+def manage_user(user, fingerprint):
+    auth_keys = user_exists(user)
```

Comment:
Overloading the testing for existence and getting the path to the local keyfile is confusing for me. And the various branches are a bit hard to follow too. Could we do something like:

```py
if not user_exists(user):
    create_user(user)

local_keys = get_local_keys(user)
github_key = get_github_key(user, fingerprint)

if github_key:
    write_file(local_keys, github_key)
else:
    write_file(local_keys, "")
```

Reply:
I don't want to create a new user until **after** we've validated they have a valid key, which is why we delay it. I'll add a comment to that effect, but I think the reason is valid?

Reply:
Regards overloading, it was just a bit ugly separating those steps out, but I take your point, I'll see if I can make in cleaner

Reply:
Right -- I see what you mean about not creating the user up front.

I think, then, that we need to be really explicit about what's going on in each of those branches. Either lots of comments or lots of carefully-named variables and helper functions.

Reply:
I've extracted a fair bit of the uncertainty into a User class, which I think now makes that code read better. Added comments too